### DISCUSSION TITLE
Module interfaces/dynamic dispatching?

### DISCUSSION THREAD

**Discussion by pshirshov** [flix/flix#6322] [Q&A]
I'm trying to figure out if the language supports dynamic dispatching in any form. Could you give me a hint if it does please?

Is it possible to declare module interfaces and then swap implementations?

If it DD is not supported, could you tell me if there are any plans for that?

**Comment by magnus-madsen** [comment]
What do you mean by dynamic dispatch?

**Reply by pshirshov** [reply]
Essentially, I would like to be able to declare module interfaces and swap dynamically bind module implementations to interfaces.

**Reply by magnus-madsen** [reply]
I see. Flix has no classes and objects, so there you would need to swap out closures or the like.

**Reply by stephentetley** [reply]
Could you use records of functions? 

This pattern is used in Haskell for `TokenParser` in Daan Leijen's original Parsec. I don't know if `TokenParser` survives in modern versions of Parsec like MegaParsec.

Flix doesn't have Rank-2 types (`forall`) so the type of a "module-as-record" will be more cumbersome than it would be in Haskell.

Tim Sheard wrote a couple of papers about this technique:

Generic Unification via Two-Level Types and Parameterized Modules
Functional Pearl: Two-Level Types and Parameterized Modules

http://web.cecs.pdx.edu/~sheard/

**Reply by pshirshov** [reply]
Well, I might invent some custom encoding but I won't be able to convince everyone to use it. 

Only a natively supported feature would work as the great unifier for all the libraries out there.

Do you think there might be a chance for this to be introduced?

**Reply by magnus-madsen** [reply]
Unlikely; but algebraic effects and handlers might suit your purpose.
